Special repository for ST clients (#29)
strimzi/proposals: 021-special-repository-for-st-clients-based-on-example-clients.md

# Special repository for ST clients based on example clients

This proposal suggests creating a new repository for `Strimzi` ST client based on 
[Strimzi client-examples](https://github.com/strimzi/client-examples).

## Current situation

Currently, we are using two clients in our STs:

>  - `InternalKafkaClients` (based on `test-client` image) 

Reviewer: I think you could describe briefly `test-client` and why it's not sufficient/you want to replace it

 - `example clients` (from [Strimzi client-examples](https://github.com/strimzi/client-examples)).

The plan is to remove the `InternalKafkaClients` and keep only `example clients`.

## Motivation

While testing `Strimzi` we need, in some cases, special configuration of clients, which is not implemented in the
`client-examples`.
How we discussed earlier, the `client-examples` repository should be really _exemplary_, and we should not add any extra
_configuration_ or _extensions_ to it. For this kind of enhancements we should have repository, which will have
`client-examples` as base, and we will be able to add special setting without disrupting the basic idea of example
clients.

> ## Proposal

Reviewer: Maybe you could put there more details? Like repo name, etc. Maybe also suggest component owners?

 * Create a new repository for `systemtest client`
    * complex implementation of clients for testing
    * we'll use both Kafka and Bridge clients from `client-examples`
    * will be based on [Strimzi client-examples](https://github.com/strimzi/client-examples) - we'll copy the
      `client-examples` code and then modify it - each repo will then _go their own way_
    * we'll be able to modify it with our special configuration
    * the main idea of example clients remain intact
    
 * The original `client-examples` repository will be kept

## Advantages

There are many things we can implement. 
Good example is returning exceptions and return codes into the `job` status 
(as we are using `k8s` jobs for deploying the example clients) and asserting it in tests - we have to grep exceptions
from the job log at the moment - which can be a problem. 

## Images and releases

The images will be built as in `client-examples` after each merged PR and pushed to `strimzi` repository on `quay.io`.
I would keep the `latest` tag, no need for extra releases and versioning.

## Kafka version

> ST client will support and use the latest Kafka version supported by `Strimzi`.

Reviewer: So if we will do Strimzi 0.22.2 and master will use Kafka 2.8, `systemtest-client` as well, how you will use `latest version supported by Strimzi` without additional versioning for `systemtest-client` ?

Reviewer 2: I read this to mean it will always support a Kafka version that's at least as recent as that supported by the operator. There's no reason why it can't be more recent. And, there's actually some value in being able to test old clients against the new broker (which you could do by having multiple jars in the image and selecting them according to an env var).

# Special repository for ST clients based on example clients

This proposal suggests creating a new repository for `Strimzi` ST client based on [Strimzi client-examples](https://github.com/strimzi/client-examples).

## Current situation

Currently, we are using two clients in our STs:
 - `InternalKafkaClients` (based on `test-client` image) 
 - `example clients` (from [Strimzi client-examples](https://github.com/strimzi/client-examples)).

The plan is to remove the `InternalKafkaClients` and keep only `example clients`. 
The `test-client` is not sufficient anymore, it can create single producer/consumer, send/receive messages, and then we can assert result.
We are stuck here, and we need to wait until a producer/consumer is finished. 
With `example clients` we are able to do a lot more.
For example - create a continuous job for sending messages with delay, 
_stack_ the producers to create a _traffic_,  add extra configuration, use different types of producer/consumer (for Bridge, Kafka, ...) and many more.  

## Motivation

While testing `Strimzi` we need, in some cases, special configuration of clients, which is not implemented in the `client-examples`.

> As we discussed offline, the `client-examples` repository should be really _exemplary_, 

Reviewer 2: I don't like to see mention of offline discussion: It suggests things are happening in private, for unknown reasons. I understand that sometimes offline discussion just happens in a non-public forum by accident. In those cases the discussion and outcome should be summarised publicly somewhere for anyone else to contribute. 

In this case I've no idea who "we" means, and in any case I don't think it matters: It is enough to say that client-examples should be exemplary, and that a dedicated test client is better than pushing the example clients in a direction that makes them less useful as examples.

Author: Good point, thanks! :)

> How we discussed earlier, the `client-examples` repository should be really _exemplary_, 

Reviewer: ```suggestion
As we discussed offline, the `client-examples` repository should be really _exemplary_, 
```

    * we'll be able to modify it with our special configuration
    * the main idea of example clients remain intact
    
 * The original `client-examples` repository will be kept

## Advantages

There are many things we can implement. 
Good example is returning exceptions and return codes into the `job` status (as we are using `k8s` jobs for deploying the example clients) and asserting it in tests. 
Currently, we have to grep exceptions from the job log - which can be a problem. 

## Images and releases

The images will be built as in `client-examples` after each merged PR and pushed to `strimzi` repository on `quay.io` with `latest` tag.
The `systemtest client` will have its own release cadence and versioning, which will depend on new features.  

## Kafka version

The `systemtest client` will support latest released version of Kafka with several older.
We'll start with version `2.5.0`, as it's the oldest supported version by latest released version of Strimzi (0.22.1),
and we'll support it until `test-clients` features will allow us to use that specific version or until we will decide to completely deprecate it.

> List of all supported versions will be present in `supported-kafka-version.yaml`.

Reviewer 3: Not sure it is that easy ... maybe you should remove this and leave it more open to how it will be done or add more details.

> For each supported Kafka version will be separate image - pattern will be `test-clients/TEST_CLIENTS_VERSION-kafka-KAFKA_VERSION`.

Reviewer 3: ```suggestion
For each supported Kafka version will be separate tag - pattern will be `test-client:TEST_CLIENTS_VERSION-kafka-KAFKA_VERSION`.
```

The plan is to remove the `InternalKafkaClients` and keep only `example clients`. The `test-client` is not sufficient
anymore, it can create single producer/consumer, send/receive messages and then we can assert result - we are 
stuck here, and we need to wait until a producer/consumer is finished. With `example clients` we are able to do 
a lot more - create a continuous job for sending messages with delay, _stack_ the producers to 
create a _traffic_,  add extra configuration, use different types of producer/consumer 
(for Bridge, Kafka, ...) and many more.  

## Motivation

While testing `Strimzi` we need, in some cases, special configuration of clients, which is not implemented in the
`client-examples`.
How we discussed earlier, the `client-examples` repository should be really _exemplary_, and we should not add any extra
_configuration_ or _extensions_ to it. For this kind of enhancements we should have repository, which will have
`client-examples` as base, and we will be able to add special setting without disrupting the basic idea of example
clients.

## Proposal

 * Create a new repository for `systemtest client`

>     * name could be `strimzi-systemtest-client`

Reviewer 3: Maybe we could just name it `test-clients`?

>     * component owners can be same as for STs

Reviewer 2: ```suggestion
    * component owners will be same as for STs
```

    * PR checks:
       * DCO

>        * build - `mvn` build, checkstyle (maybe with some simple UTs or ITs)

Reviewer 2: ```suggestion
       * build - `mvn` build, checkstyle (with some simple UTs or ITs)
```

    * we'll use both Kafka and Bridge clients from `client-examples`
    * will be based on [Strimzi client-examples](https://github.com/strimzi/client-examples) - we'll copy the
      `client-examples` code and then modify it - each repo will then _go their own way_
    * we'll be able to modify it with our special configuration
    * the main idea of example clients remain intact
    
 * The original `client-examples` repository will be kept

## Advantages

There are many things we can implement. 
Good example is returning exceptions and return codes into the `job` status 
(as we are using `k8s` jobs for deploying the example clients) and asserting it in tests - we have to grep exceptions
from the job log at the moment - which can be a problem. 

## Images and releases

The images will be built as in `client-examples` after each merged PR and pushed to `strimzi` repository on `quay.io`
with `latest` tag.
Because we can use (or will use) different versions of Kafka for each Strimzi release, we should release the new version
of `systemtest-client` together with Strimzi - to prevent issues with features added for higher version of Kafka, to

> make it easier to handle it in STs and make it less confusing.

Reviewer 2: If it's coupled tightly to the Strimzi operators ST then it should be in that repo and thus automatically always in sync. If you want it in a separate repo for use with the bridge, or other repos, then you should aim to have a versioned API (of some sort) and it's own release cadence (which you can chose based on new features, of course).

> The image will contain `.jars` of all supported Kafka versions, the ones for specified version will be used.

Reviewer 3: How would you implement this?

If the user specifies unsupported Kafka version, error message will appear in _status_ and _log_ together with list of supported versions.
List of all supported versions will be present in `supported-kafka-version.yaml`.

> We'll start with version `2.5.0`, as it's the oldest supported version by Strimzi,

Reviewer 3: main branch now uses only 2.6.0+.

Author: This is meant for latest released Strimzi - `0.22.1` - where the `2.5.0` is supported, maybe I should specify it.
